### PR TITLE
Organization Related Application Deletion

### PR DESCRIPTION
**Pull request by rashm1n**
$ Subject

## Purpose

Delete application data related to a certain organization

### Related Tasks

wso2-enterprise/choreo#6643

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeSQLConstants.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package org.wso2.carbon.apimgt.cleanup.service;
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeSQLConstants.java:1]: License header?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java [hunk 2]
```diff
@@ -1004,7 +1004,7 @@
             "   AM_SUBSCRIBER SUB " +
             " WHERE " +
             "   LOWER(APP.NAME) = LOWER(?)" + "   " +
-            "   AND ORGANIZATION = ? " +
+            "   AND APP.ORGANIZATION = ? " +
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java:1007]: Shouldn’t this change be revert as we will remove the subscriber deletion flow?

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java [hunk 3]
```diff
@@ -15,16 +15,115 @@
...
+    }
+
+    public ApplicationPurge(String username) {
+        this.username = username;
+        organizationPurgeDAO = OrganizationPurgeDAO.getInstance();
+        initTaskList();
+    }
+
+    public ApplicationPurge(OrganizationPurgeDAO organizationPurgeDAO, String username) {
+        this.organizationPurgeDAO = organizationPurgeDAO;
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java:56]: call as this and initialize upper constructor

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class OrganizationPurgeDAO {
+
+    private static final Log log = LogFactory.getLog(ApiMgtDAO.class);
+    private static OrganizationPurgeDAO INSTANCE = null;
+    private boolean multiGroupAppSharingEnabled = false;
+
+    public OrganizationPurgeDAO() {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:47]: make private

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+     *
+     * @param organization Organization
+     * @throws APIManagementException if failed to remove pending subscriptions
+     */
+    public void removePendingSubscriptions(String organization) throws APIManagementException {
+
+        String query = OrganizationPurgeSQLConstants.DELETE_PENDING_SUBSCRIPTIONS;
+
+        try (Connection connection = APIMgtDBUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:78]: before creating prepared statement need to make connection autocommit false

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+
+            connection.setAutoCommit(false);
+
+            preparedStatement.setString(1, organization);
+
+            preparedStatement.executeUpdate();
+            connection.commit();
+
+        } catch (SQLException e) {
+            String msg = "Error occurred while removing pending subscriptions for organization: "+organization;
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:88]: handle rollback

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+        }
+    }
+
+    /**
+     * Remove application creation workflows
+     *
+     * @param organization Organization
+     * @throws APIManagementException if failed to remove application creation workflows
+     */
+    public void removeApplicationCreationWorkflows(String organization)
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:99]: same goes here as well

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+     * Remove pending Application Registrations
+     *
+     * @param organization Organization
+     * @throws APIManagementException when failed to delete pending application registrations
+     */
+    public void deletePendingApplicationRegistrations(String organization) throws APIManagementException {
+        String query = OrganizationPurgeSQLConstants.REMOVE_PENDING_APPLICATION_REGISTRATIONS;
+
+        try (Connection connection = APIMgtDBUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:129]: need autocommit false

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        PreparedStatement deleteApp = null;
+        ResultSet rs = null;
+
+        String getConsumerKeyQuery = OrganizationPurgeSQLConstants.GET_CONSUMER_KEYS_OF_APPLICATION_LIST_SQL;
+
+        String deleteDomainAppQuery = SQLConstants.REMOVE_APPLICATION_FROM_DOMAIN_MAPPINGS_SQL;
+
+        String deleteApplicationQuery = OrganizationPurgeSQLConstants.REMOVE_APPLICATION_LIST_FROM_APPLICATIONS_SQL;
+
+        try {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:160]: re-write this to use try-resource functionality

**Review by tharindu1st — CHANGES_REQUESTED** [review]

**Review by HiranyaKavishani — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java (hunk 3 above)
Comment (HiranyaKavishani) [reply components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java:57]: @rashm1n  Could you please fix this?

Comment (rashm1n) [reply components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java:57]: @HiranyaKavishani this is already fixed. Resolving.

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java [hunk 10]
```diff
@@ -15,16 +15,114 @@
...
  * under the License.
  */
+
 package org.wso2.carbon.apimgt.cleanup.service;
 
+import com.google.gson.Gson;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.apimgt.api.APIManagementException;
+
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java:25]: Remove unnecessary blank lines

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java [hunk 11]
```diff
@@ -15,16 +15,114 @@
...
+                    log.error("Error while deleting Application Data in organization " + organization, e);
+                    applicationPurgeTaskMap.put(task.getKey(), APIConstants.OrganizationDeletion.FAIL);
+                    log.info("Re-trying to execute " + task.getKey() + " process for organization" + organization, e);
+
+                    if (++count == maxTries) {
+                        log.error("Cannot execute " + task.getKey() + " process for organization" + organization, e);
+                        applicationPurgeTaskMap.put(task.getKey(), e.getMessage());
+                        break;
+                    }
+
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java:98]: Remove blank line

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java [hunk 12]
```diff
@@ -15,16 +15,114 @@
...
+import org.wso2.carbon.apimgt.impl.APIConstants;
+import org.wso2.carbon.apimgt.impl.utils.APIUtil;
+
 import java.util.LinkedHashMap;
+import java.util.Map;
 
 public class ApplicationPurge implements OrganizationPurge {
+    private String username;
+    protected OrganizationPurgeDAO organizationPurgeDAO;
+    private static final Log log = LogFactory.getLog(ApplicationPurge.class);
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/ApplicationPurge.java:35]: Use `static final` statements first

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 13]
```diff
@@ -206,6 +212,245 @@ public void deleteKeyManagerConfigurationList(List<KeyManagerConfigurationDTO> k
...
     }
 
+    /**
+     * Remove pending subscriptions for a given application id list
+     *
+     * @param organization Organization
+     * @throws APIManagementException if failed to remove pending subscriptions
+     */
+    public void removePendingSubscriptions(String organization) throws APIManagementException {
+
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:222]: Remove blank line

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 14]
```diff
@@ -206,6 +212,245 @@ public void deleteKeyManagerConfigurationList(List<KeyManagerConfigurationDTO> k
...
+        String query = OrganizationPurgeSQLConstants.DELETE_PENDING_SUBSCRIPTIONS;
+
+        try (Connection connection = APIMgtDBUtil.getConnection()) {
+            connection.setAutoCommit(false);
+
+            try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+                preparedStatement.setString(1, organization);
+
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:230]: Remove blank line

Comment (lasanthaS) [reply components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:230]: Do the same for other places as well.

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 15]
```diff
@@ -206,6 +212,245 @@ public void deleteKeyManagerConfigurationList(List<KeyManagerConfigurationDTO> k
...
+    /**
+     * Deletes Applications along with subscriptions, keys and registration data
+     *
+     * @param organization Organization
+     * @throws APIManagementException if failed to delete applications for organization
+     */
+    public void deleteApplicationList(String organization) throws APIManagementException {
+
+        String getConsumerKeyQuery = OrganizationPurgeSQLConstants.GET_CONSUMER_KEYS_OF_APPLICATION_LIST_SQL;
+
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:321]: Don't put blank lines in these places. Only place if there is a logical differences

File: components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java [hunk 16]
```diff
@@ -206,6 +212,245 @@ public void deleteKeyManagerConfigurationList(List<KeyManagerConfigurationDTO> k
...
+            }
+
+            try (PreparedStatement prepStmtGetConsumerKey = connection.prepareStatement(getConsumerKeyQuery);
+                    PreparedStatement deleteDomainApp = connection.prepareStatement(deleteDomainAppQuery)) {
+                prepStmtGetConsumerKey.setString(1, organization);
+
+                try (ResultSet rs = prepStmtGetConsumerKey.executeQuery()) {
+                    while (rs.next()) {
+                        String consumerKey = rs.getString(APIConstants.FIELD_CONSUMER_KEY);
+                        String keyManagerName = rs.getString("NAME");
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.cleanup.service/src/main/java/org/wso2/carbon/apimgt/cleanup/service/OrganizationPurgeDAO.java:340]: Use constants

**Pull request merged**